app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java

```diff
@@ -0,0 +1,519 @@
+    public static final int METHOD_RECORD = 200_003;
+    /**
+     * Tell devices to remove video recording preparation.
+     */
+    public static final int METHOD_STOP_PREPARE = 200_004;
+
+    private long mUpcomingTriggerTimeNs;
+
+    /**
+     * Constructor passed in with: - context - For setting UI elements and triggering captures. -
+     * captureButton - The button used to send at trigger request by the leader. - statusView - The
+     * TextView used to show currently connected clients on the leader device.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    public SoftwareSyncController(
+            MainActivity mainActivity, PhaseAlignController phaseAlignController, PeriodCalculator periodCalculator) {
+        mMainActivity = mainActivity;
+        mPhaseAlignController = phaseAlignController;
+        mPeriodCalculator = periodCalculator;
+        mSoftwareSyncHelper = new SoftwareSyncHelper(mainActivity, this);
+
+        setupSoftwareSync();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    private void setupSoftwareSync() {
+        Log.w(TAG, "setup SoftwareSync");
+        if (mSoftwareSync != null) return;
+
+        // Get Wifi Manager and use NetworkHelpers to determine local and leader IP addresses.
+        WifiManager wifiManager = (WifiManager) mMainActivity.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
+        InetAddress leaderAddress;
+        InetAddress localAddress;
+
+        // Use last 4 digits of the serial as the name of the client.
+        String name = lastFourSerial();
+        Log.w(TAG, "Name/Serial# (Last 4 digits): " + name);
+
+        // Determine leadership.
+        try {
+            NetworkHelpers networkHelper = new NetworkHelpers(wifiManager);
+            mIsLeader = networkHelper.isLeader();
+
+            if (mIsLeader) {
+                // IP determination is not yet implemented for a leader.
+                localAddress = null;
+                leaderAddress = null;
+            } else {
+                localAddress = networkHelper.getIPAddress();
+                leaderAddress = networkHelper.getHotspotServerAddress();
+            }
+
+            Log.w(
+                    TAG,
+                    String.format(
+                            "Current IP: %s , Leader IP: %s | Leader? %s",
+                            localAddress, leaderAddress, mIsLeader ? "Y" : "N"));
+        } catch (SocketException e) {
+            Log.e(TAG, "Error: " + e);
+            throw new IllegalStateException(
+                    "Unable to determine leadership, check if WiFi or hotspot is enabled.", e);
+        } catch (UnknownHostException e) {
+            Log.e(TAG, "Error: " + e);
+            throw new IllegalStateException(
+                    "Unable to get IP addresses, check Network permissions.", e);
+        }
+
+        // Set up shared rpcs.
+        Map<Integer, RpcCallback> sharedRpcs = new HashMap<>();
+
+        //sharedRpcs.put(
```

Comment:
I think we don't need this commented code, nothing important left here

Reply:
Removed this, but left the corresponding `METHOD_SET_TRIGGER_TIME` field commented so that its int code won't be used in the future by some other RPC method (just in case we care about backwards compatibility)